src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/UpToDateCheckImplicitConfiguredInput.cs

```diff
@@ -220,17 +236,34 @@ private UpToDateCheckImplicitConfiguredInput(ProjectConfiguration projectConfigu
             ItemHash = itemHash;
 
             var setNames = new HashSet<string>(BuildUpToDateCheck.SetNameComparer);
-            AddKeys(upToDateCheckInputItemsByKindBySetName);
-            AddKeys(upToDateCheckOutputItemsByKindBySetName);
-            AddKeys(upToDateCheckBuiltItemsByKindBySetName);
+            AddSetNames(upToDateCheckInputItemsByKindBySetName);
+            AddSetNames(upToDateCheckOutputItemsByKindBySetName);
+            AddSetNames(upToDateCheckBuiltItemsByKindBySetName);
             setNames.Remove(BuildUpToDateCheck.DefaultSetName);
             SetNames = setNames.OrderBy(n => n, BuildUpToDateCheck.SetNameComparer).ToImmutableArray();
 
-            void AddKeys(ImmutableDictionary<string, ImmutableDictionary<string, ImmutableArray<string>>> dictionary)
+            var kindNames = new HashSet<string>(BuildUpToDateCheck.KindNameComparer);
+            AddKindNames(upToDateCheckInputItemsByKindBySetName);
+            AddKindNames(upToDateCheckOutputItemsByKindBySetName);
+            AddKindNames(upToDateCheckBuiltItemsByKindBySetName);
+            KindNames = kindNames.OrderBy(n => n, BuildUpToDateCheck.KindNameComparer).ToImmutableArray();
```

Comment:
Here we specify the order of kind names as alphabetical (similar to how we treated "set" names previously).